packages/api-docs-builder/utils/getPropsFromComponentNode.ts

```diff
@@ -0,0 +1,345 @@
+  parsedComponents.forEach((parsedComponent) => {
+    Object.keys(parsedComponent.props).forEach((propName) => {
+      if (!squashedProps[propName]) {
+        squashedProps[propName] = parsedComponent.props[propName];
+      } else {
+        squashedProps[propName].signatures = [
+          ...squashedProps[propName].signatures,
+          ...parsedComponent.props[propName].signatures,
+        ];
+      }
+    });
+  });
+
+  const squashedParsedComponent: ParsedComponent = {
+    ...parsedComponents[0],
+    props: squashedProps,
+  };
+
+  Object.keys(squashedParsedComponent.props).forEach((propName) => {
+    squashedParsedComponent.props[propName].onlyUsedInSomeSignatures =
+      squashedParsedComponent.props[propName].signatures.length < signatures.length;
+  });
+
+  return squashedParsedComponent;
+}
+
+export interface GetPropsFromComponentDeclarationOptions {
+  project: TypeScriptProject;
+  node: ts.Node;
+  /**
+   * Called before a PropType is added to a component/object
+   * @returns true to include the prop, false to skip it
+   */
+  shouldInclude?: (data: { name: string; depth: number }) => boolean;
+  /**
+   * Control if const declarations should be checked
+   * @default false
+   * @example declare const Component: React.JSXElementConstructor<Props>;
+   */
+  checkDeclarations?: boolean;
+}
+
+function getPropsFromVariableDeclaration({
+  node,
+  project,
+  checkDeclarations,
+  shouldInclude,
+}: { node: ts.VariableDeclaration } & Pick<
+  GetPropsFromComponentDeclarationOptions,
+  'project' | 'checkDeclarations' | 'shouldInclude'
+>) {
+  const type = project.checker.getTypeAtLocation(node.name);
+  if (!node.initializer) {
+    if (
+      checkDeclarations &&
+      type.aliasSymbol &&
+      type.aliasTypeArguments &&
+      project.checker.getFullyQualifiedName(type.aliasSymbol) === 'React.JSXElementConstructor'
+    ) {
+      const propsType = type.aliasTypeArguments[0];
+      if (propsType === undefined) {
+        throw new TypeError(
+          'Unable to find symbol for `props`. This is a bug in typescript-to-proptypes.',
+        );
+      }
+      return parsePropsType({
+        name: node.name.getText(),
+        type: propsType,
+        location: node.name,
+        shouldInclude,
+        sourceFile: node.getSourceFile(),
+      });
+    }
+
+    if (checkDeclarations) {
+      return parseFunctionComponent({
+        node,
+        shouldInclude,
+        project,
+      });
+    }
+
+    return null;
+  }
+
+  if (
+    (ts.isArrowFunction(node.initializer) || ts.isFunctionExpression(node.initializer)) &&
+    node.initializer.parameters.length === 1
+  ) {
+    return parseFunctionComponent({
+      node,
+      shouldInclude,
+      project,
+    });
+  }
+  //  x = React.memo((props:type) { return <div/> })
+  //  x = React.forwardRef((props:type) { return <div/> })
+  if (ts.isCallExpression(node.initializer) && node.initializer.arguments.length > 0) {
```

Comment:
We also have `React.memo(React.forwardRef(...))` (Base UI's Option), but since there are no changes in generated files, I assume it works fine :+1:

Reply:
I guess it does :ok_hand:

Reply:
@flaviendelangle, it looks like it doesn't after all. While working on https://github.com/mui/material-ui/pull/36287, I changed the wording of the PropTypes warning message and noticed that it was not updated on some components.
The affected components are:

- base/Option
- base/Portal
- base/TextareaAutosize
- joy/Box
- lab/Timeline
- material-next/SliderRoot
- material-next/SliderRail
- material-next/SliderTrack
- material-next/SliderThumb
- material-next/SliderValueLabel
- material-next/SliderMark
- material-next/SliderMarkLabel
- material/Popper
- material/SliderRoot
- material/SliderRail
- material/SliderTrack
- material/SliderThumb
- material/SliderValueLabel
- material/SliderMark
- material/SliderMarkLabel

Reply:
Do we agree that the previous version did not handle this scenario as well ?
If so, I can have a look on how to improve the new one :+1:

Reply:
Partially. For example, base/Option did not work correctly before, but joy/Box was fine.

Reply:
Do you know why there is no diff in this PR for joy/Box then?

Reply:
For base/Option, the injector did not handler nested wrapper function list `memo(forwardRef(...))`, I have a working fix.

For joy/Box, I'm investigating

Reply:
For joy/Box, I was returning too soon so all the `createX` were probably broken at the parsing phase

Reply:
For base/Portal (and probably others) we don't support correctly the `function() {} as XXX`

Reply:
For `SliderRoot` and equivalent, I suspect that the proptypes have been written manually, because I see nothing in the code to generate them for this kind of syntax.